w3-blob.md

```diff
@@ -0,0 +1,199 @@
+
+## Abstract
+
+W3 blob protocol allows authorized agents to store arbitrary content blobs with a storage provider.
+
+## Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+# Introduction
+
+W3 blob protocol provides core building block for storing content and sharing access to it through UCAN authorization system. It is successor to the [store protocol] which no longer requires use of [Content Archive][CAR]s even if in practice clients can continue to use it for storing shards of large DAGs.
+
+## Concepts
+
+### Roles
+
+There are several distinct roles that [principal]s may assume in this specification:
+
+| Name        | Description                                                                                                                                    |
+| ----------- | ---------------------------------------------------------------------------------------------------------------------------------------------- |
+| Principal | The general class of entities that interact with a UCAN. Identified by a DID that can be used in the `iss` or `aud` field of a UCAN. |
+| Agent       | A [Principal] identified by [`did:key`] identifier, representing a user in an application. |
+| Issuer | A [principal] delegating capabilities to another [principal]. It is the signer of the [UCAN]. Specified in the `iss` field of a UCAN. |
+| Audience | Principal access is shared with. Specified in the `aud` field of a UCAN. |
+
+
+### Space
+
+A namespace, often referred as a "space", is an owned resource that can be shared. It corresponds to a unique asymmetric cryptographic keypair and is identified by a [`did:key`] URI.
+
+### Blob
+
+Blob is a fixed size byte array addressed by the [multihash]. Usually blobs are used to represent set of IPLD blocks at different byte ranges.
+
+# Capabilities
+
+## Add Blob
+
+Authorized agent MAY invoke `/space/content/add/blob` capability on the [space] subject to store specific byte array.
+
+> Note that storing a blob does not imply advertising it on the network or making it publicly available.
+
+### Add Blob Capability
+
+#### Add Blob Capability Schema
+
+```ipldsch
+type struct AddBlob {
+  cmd   "/space/content/add/blob"
+  sub   SpaceDID
+  args  Blob
+}
+
+type struct Blob {
+  content   Multihash
+  size      Int
+}
+```
+
+#### Blob Content
+
+Blob `content` field MUST be a [multihash] digest of the blob payload bytes, uniquely identifying blob.
+
+#### Blob Size
+
+Blob `size` field MUST be set to the size of the blob in bytes.
+
+### Add Blob Receipt
+
+#### Add Blob Result
+
+```ipldsch
+type BlobAddResult union {
+  BlobAddSuccess    "ok"
+  BlobAddFailure    "error"
+} representation keyed
+
+type BlobAddSuccess union {
+  BlobAddAllocation     "allocated"
+  &UCANLocationClaim    "committed"
```

Comment:
Is this for returning when we already have content? Otherwise it should be async right?

in other words, the `blob/close` kind of capability executed when we see that upload made it to write target

Reply:
Yeah this was basically what we have right now in the store/* API, although I do think you might be raising a good point here perhaps we need same result and this should be linked through an effect.

It might help to go over several options so we can compare tradeoffs and choose a right compromise.

Reply:
TODO: How do we deal with /add → /remove → /add if we say we issue the same receipt.

Reply:
```js
{
  ran: { "/": "bafy..add" },
  out: { ok: { allocated: 123 } },
  next: [
    { cmd: "service/blob/wait", with: "did:web:web3.storage", args: { content: 'hash', size: 123 }  },
    { cmd: "assert/location", with: "did:web:web3.storage", args: { url: "https://...", headers: {}, _: { "await/ok": {"/": "bafy...blobawait" } } } } 
  ]

}
```

Reply:
```js
{
  iss: "did:web:web3.storage",
 aud: did:key:alice
  att: [{
     can: assert/location
      with: "did:web:web3.storage",
      nb: {
          content: {},
          url: ...
          headers: ...
          range: [0, 123]
      }
  }]
}
```

Reply:
did:web:web3.storage → (service key, space did) → did:key:alice